### DISCUSSION TITLE
Use a GitHub App instead of a personal access token

### DISCUSSION THREAD

**Discussion by agaudreault** [renovatebot/renovate#10026] [Request Help] [answered]
We are using a custom user for our renovate self hosted installation with GitHub Cloud, however we are hitting rate-limit.

- A user have 5000/request per hour on the GitHub api.
- A GitHub App can have up to 15,000/request per hour.

Since renovate can be installed on GitHub with the Official app, is there a way to configure it with a private GitHub app (using `x-access-token`) instead of a personal access token ?

From the code, it seems like we can get passed the initial "user discovery" if we configure a `username` and a `gitAuthor` in the configuration. However, the codes use APIs not available to GitHub apps later on during the autodiscovery and returns a `Resource not accessible by integration` error.

- `/user`: https://github.com/renovatebot/renovate/blob/757de651fc25603eb6595d2a3f6ace74445bc6b7/lib/platform/github/user.ts#L22 could be `/app`, but can be ignore with `username` specified.
- '/user/emails': https://github.com/renovatebot/renovate/blob/757de651fc25603eb6595d2a3f6ace74445bc6b7/lib/platform/github/user.ts#L48 has no equivalent, but can be ignored with `gitAuthor` specified.
- `/user/repos`: https://github.com/renovatebot/renovate/blob/757de651fc25603eb6595d2a3f6ace74445bc6b7/lib/platform/github/index.ts#L120 and https://github.com/renovatebot/renovate/blob/757de651fc25603eb6595d2a3f6ace74445bc6b7/lib/platform/github/index.ts#L296 could be `/app/installations`

```
"message": "Response code 403 (Forbidden)",
"stack": "HTTPError: Response code 403 (Forbidden)\n    at Request.<anonymous> (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)",
"options": {
  "headers": {
    "user-agent": "https://github.com/renovatebot/renovate",
    "accept-encoding": "gzip, deflate",
    "accept": "application/vnd.github.machine-man-preview+json",
    "authorization": "***********"
  },
  "url": "https://api.github.com/user/repos?per_page=100",
  "username": "",
  "password": "",
  "method": "GET",
  "http2": false
},
"response": {
  "statusCode": 403,
  "statusMessage": "Forbidden",
  "body": {
    "message": "Resource not accessible by integration",
    "documentation_url": "https://docs.github.com/rest/reference/repos#list-repositories-for-the-authenticated-user"
  },
  [...]
  ERROR: GitHub getRepos error
"err": {
  "message": "integration-unauthorized",
  "stack": "Error: integration-unauthorized\n    at handleGotError (/usr/src/app/node_modules/renovate/lib/util/http/github.ts:90:11)\n    at GithubHttp.request (/usr/src/app/node_modules/renovate/lib/util/http/github.ts:218:7)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)\n    at GithubHttp.requestJson (/usr/src/app/node_modules/renovate/lib/util/http/index.ts:180:17)\n    at Proxy.getRepos (/usr/src/app/node_modules/renovate/lib/platform/github/index.ts:120:17)\n    at Object.autodiscoverRepositories (/usr/src/app/node_modules/renovate/lib/workers/global/autodiscover.ts:24:20)\n    at Object.start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:86:14)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:16:22"
}
```
Renovate version: 24.114.2

For more details, GitHub App endpoints are available in https://docs.github.com/en/rest/overview/endpoints-available-for-github-apps

**Comment by viceice** [comment]
There are some users successfully using renovate with a custom app token. So yes, you need to configure `username  and `gitAuthor`.

I think you can't use autodiscover with an app token, so you need to manually configure the repo list. You generate it before running renovate and pass it to renovate via available [repositories](https://docs.renovatebot.com/self-hosted-configuration/#repositories) option.

**Reply by viceice** [reply]
Checkout #6371 too

**Reply by viceice** [reply]
@rarkins @HonkingGoose  Maybe we should add some docs update [here](https://docs.renovatebot.com/self-hosting/#authentication)  to include github app token usage?

**Reply by rarkins** [reply]
You need to:
- Prefix the token with `x-access-token:`
- Configure `username` (e.g. `myorgrenovate[bot]`)
- Configure `gitAuthor` (e.g. `myorgrenovate[bot]@noreply.users.github.com`)
- Configure the list of repositories manually

**Reply by agaudreault** [reply]
Alright so I might create a feature request and a PR to make the auto discovery available to GitHub App. We generate quite a big regex to find all our projects so this is very useful.

**Reply by HonkingGoose** [reply]
> @rarkins @HonkingGoose Maybe we should add some docs update [here](https://docs.renovatebot.com/self-hosting/#authentication) to include github app token usage?

Documenting things fully is always good. 😄 

@rarkins Are you willing to write a draft for the docs update? I don't know enough about this to make a meaningful draft. I can review your draft of course. 😉

**Reply by rarkins** [reply]
I can't prioritize this right now, because it's not applicable to a wide number of users

**Reply by HonkingGoose** [reply]
Do you want me to make a low priority issue for this then, so that we at least capture the improvement idea?

**Reply by HonkingGoose** [reply]
See #10040 for the low priority issue about adding the docs.

**Comment by agaudreault** [comment, marked as answer]
Fixed in https://github.com/renovatebot/renovate/pull/13406
